ENH: raise ValueError if invalid period freq pass to asfreq when the index of df is a PeriodIndex

natmokval opened the pull request:
xref #55785, #52064

`PeriodIndex.asfreq` silently converts for offsets such as `offsets.MonthBegin()`, offsets.BusinessMonthEnd(), etc. (with no attribute '`_period_dtype_code'`) frequency to period frequency (in this case `'M'`).

Reproducible Example:
```
>>> import pandas as pd
>>> from pandas.tseries import offsets
>>>
>>> index = pd.PeriodIndex(["2020-01-01", "2021-01-01"], freq="Q")
>>> index.asfreq(freq=offsets.MonthBegin(1))
PeriodIndex(['2020-03', '2021-03'], dtype='period[M]')
>>>
>>> index.asfreq(freq=offsets.BusinessMonthEnd(1)) 
PeriodIndex(['2020-03', '2021-03'], dtype='period[M]')
```
the correct behaviour would be raising an Error:
```
>>> index.asfreq(freq=offsets.MonthBegin(1))
ValueError: <MonthBegin> is not supported as period frequency
```

Another problem: so far in the example below
```
>>> index = pd.PeriodIndex(["2020-01-01", "2021-01-01"], freq="M")
>>> index.asfreq(freq='BMS')
```
we get on main
```
AttributeError: 'pandas._libs.tslibs.offsets.BusinessMonthBegin' object has no attribute '_period_dtype_code'
```
the correct behaviour would be raising an Error:
```
ValueError: BMS is not supported as period frequency
```
added to the definition of asfreq a check if `string` denoting frequency is supported as period frequency. If the index of a DataFrame is a PeriodIndex and the frequency is invalid period frequency a `ValueError` is raised.

natmokval committed:
raise ValueError in asfreq for invalid period freq

natmokval commented:
@MarcoGorelli, could you please take a look at this PR?

MarcoGorelli commented on pandas/core/resample.py:
```diff
@@ -2827,6 +2827,16 @@ def asfreq(
         if how is None:
             how = "E"
 
+        if isinstance(freq, str):
```
a few lines down, there's ` obj.index.asfreq(freq, how=how)` - should these validation go in there?

natmokval replied:
thanks, you are right. I moved the check `if isinstance(freq, str)` to the definition of `asfreq` in `pandas/core/indexes/period.py`

natmokval committed:
Merge branch 'main' into enh-error-msg-asfreq-invalid-period-freq

natmokval committed:
correct PeriodIndex.asfreq, fix test, add tests

MarcoGorelli commented on pandas/core/indexes/period.py:
```diff
@@ -205,6 +210,18 @@ def _resolution_obj(self) -> Resolution:
+            offset = to_offset(freq, is_period=True)
+            if hasattr(offset, "_period_dtype_code"):
+                freq = freq_to_period_freqstr(offset.n, offset.name)
+            elif offset.name == freq.replace(f"{offset.n}", ""):
+                raise ValueError(
+                    f"Invalid offset: '{offset.name}' for converting time series "
+                    f"with PeriodIndex."
+                )
```
I don't really understand this, but I think you don't need it - just raise `ValueError(INVALID_FREQ_ERR_MSG.format(f"{freq}"))` in the `else` branch below?

natmokval replied:
I use two different error messages here because we raise an error of invalid freq such as `“ABC”` and also for `freq=“”BMS”` which is valid for offsets, but invalid for period. I agree, that it can be confusing, so better to use the standard error message `"Invalid frequency: {freq}"`

MarcoGorelli commented on pandas/core/indexes/period.py:
```diff
@@ -205,6 +210,18 @@ def _resolution_obj(self) -> Resolution:
+            else:
+                raise ValueError(INVALID_FREQ_ERR_MSG.format(f"{freq}"))
+
         arr = self._data.asfreq(freq, how)
```
can you follow this `asfreq` and put this validation even deeper?

natmokval replied:
thanks, I did as you suggested and moved this validation to `to_offset`. It works very well.

MarcoGorelli reviewed (changes requested):
thanks for working on this

a lot of these validation paths go through `to_offset` - could we just check that if you run `to_offset(..., is_period=True)` then it raises an informative error message if the return value doesn't have a `_period_dtype_code` attribute?

So here

https://github.com/pandas-dev/pandas/blob/02011f2b81743f4e9d100aea6c791fa5e8de9933/pandas/_libs/tslibs/offsets.pyx#L4840-L4849

instead of returning, assign the return value to a variable (say, `delta`, as that's what's used below)

then, just after

https://github.com/pandas-dev/pandas/blob/02011f2b81743f4e9d100aea6c791fa5e8de9933/pandas/_libs/tslibs/offsets.pyx#L4964-L4965

raise if `is_period` and `delta` doesn't have a `_period_dtype_code` attribute?

natmokval committed:
move raising ValueError to to_offset

natmokval committed:
fix pre-commit errors

MarcoGorelli commented on pandas/_libs/tslibs/offsets.pyx:
```diff
@@ -4846,7 +4846,7 @@ cpdef to_offset(freq, bint is_period=False):
 
     elif PyDelta_Check(freq):
-        return delta_to_tick(freq)
+        delta = delta_to_tick(freq)
```
maybe on line 4841, we can do
```
    if isinstance(freq, BaseOffset):
        delta = freq
```
so then that gets validated too?

natmokval replied:
I am unsure if it's possible. I tried to do this assignment on line 4841
```
if isinstance(freq, BaseOffset):
    delta = freq
```
but then I got failures.
The reason: if we replace `return freq` with the `delta = freq`, we go to the line 4962 and assign delta to None and then on line 4965 we raise a ValueError.
Which is why instead of the assignment `delta = freq` I added the check 
```
if is_period and not hasattr(freq, "_period_dtype_code"):
    raise ValueError(f"{freq.base} is not supported as period frequency")
```

natmokval replied:
thanks, I saw you already made commit for it

natmokval committed:
correct def to_offset, dt64arr_to_periodarr and fix tests

MarcoGorelli reviewed (commented):
nice one - does this mean it's also possible to get rid of the `except` in

https://github.com/pandas-dev/pandas/blob/acd914dedba087b3642e110aadac665006da23c0/pandas/core/resample.py#L2814-L2820

?

natmokval committed:
fix tests

natmokval commented:
> nice one - does this mean it's also possible to get rid of the `except` in
> 
> https://github.com/pandas-dev/pandas/blob/acd914dedba087b3642e110aadac665006da23c0/pandas/core/resample.py#L2814-L2820
> 
> ?

I am afraid we can't rid of this block yet. When I removed this check the `test_asfreq_invalid_period_freq` in `pandas/tests/resample/test_period_index.py` failed. 

Without this block we silently convert index, for example:
```
index = pd.PeriodIndex(["2020-01-01", "2021-01-01"], freq="Q")
df = pd.DataFrame({"a": pd.Series([0, 1], index=index)})
res = df.asfreq(freq=offsets.BusinessMonthEnd(1))
print(res.index)

PeriodIndex(['2020-03', '2021-03'], dtype='period[M]')
```

natmokval committed:
correct def to_offset, fix tests

MarcoGorelli reviewed (commented):
should be good thanks, just one minor comment

does this change anything since 2.2? if so, I'd say that a whatsnew note in 3.0 is warranted (or in 2.2.1 if it should be backported, for example if anything introduced in 2.2 wasn't correct and needs fixing here)

natmokval committed:
add a note to v2.2.1.rst

natmokval committed:
resolve conflict

natmokval commented:
> should be good thanks, just one minor comment
> 
> does this change anything since 2.2? if so, I'd say that a whatsnew note in 3.0 is warranted (or in 2.2.1 if it should be backported, for example if anything introduced in 2.2 wasn't correct and needs fixing here)

yes, I think we have an incorrect conversion in 2.2
```
>>> index = pd.PeriodIndex(["2020-01-01", "2021-01-01"], freq="M")
>>> index.asfreq(freq=offsets.MonthBegin(1))
PeriodIndex(['2020-03', '2021-03'], dtype='period[M]')
>>>
>>> index.asfreq(freq=offsets.YearBegin(1))
PeriodIndex(['2020', '2021'], dtype='period[Y-DEC]')
>>>
>>> index.asfreq(freq=offsets.BusinessMonthEnd(1))
PeriodIndex(['2020-03', '2021-03'], dtype='period[M]')
```
I fixed it in this PR and added an entry the 2.2.1. Now we raise a ValueError, e.g.
```
>>> index.asfreq(freq=offsets.MonthBegin(1))
ValueError: <MonthBegin> is not supported as period frequency
```

natmokval commented on pandas/tests/scalar/period/test_period.py:
```diff
@@ -46,15 +46,16 @@ def test_offsets_not_supported(self, freq, freq_msg):
 
     def test_invalid_frequency_error_message(self):
-        msg = "WeekOfMonth is not supported as period frequency"
-        with pytest.raises(TypeError, match=msg):
+        msg = "WOM-1MON is not supported as period frequency"
+        with pytest.raises(ValueError, match=msg):
             Period("2012-01-02", freq="WOM-1MON")
```
I am unsure if it's possible to do this assignment on line 4841
```
if isinstance(freq, BaseOffset):
    delta = freq
```
but then I got failures.
The reason: if we replace return freq with the delta = freq, we go to the line 4962 and assign delta to None and then on line 4965 we raise a ValueError.

Instead of the assignment delta = freq I added the check
```
if is_period and not hasattr(freq, "_period_dtype_code"):
    raise ValueError(f"{freq.base} is not supported as period frequency")
```

MarcoGorelli committed:
Merge remote-tracking branch 'upstream/main' into enh-error-msg-asfreq-invalid-period-freq

MarcoGorelli committed:
try simplifying / clarifying

MarcoGorelli commented on pandas/plotting/_matplotlib/timeseries.py:
```diff
@@ -205,7 +205,10 @@ def _get_ax_freq(ax: Axes):
 
 def _get_period_alias(freq: timedelta | BaseOffset | str) -> str | None:
-    freqstr = to_offset(freq, is_period=True).rule_code
+    if isinstance(freq, BaseOffset):
+        freqstr = freq.name
+    else:
+        freqstr = to_offset(freq, is_period=True).rule_code
```
ok I presume you did this because otherwise the plotting tests fail? I'd suggest leaving a comment in your PR explaining that in the future, else it's very hard to understand why this is being done

natmokval replied:
sorry, I'll leave a comment next time in my PR.

Yes, you are right, without these changes tests for plotting in `pandas/tests/plotting/test_datetimelike.py` and `pandas/tests/plotting/test_series.py` fail.

MarcoGorelli committed:
revert _get_period_alias simplification

MarcoGorelli reviewed (approved):
wonderful, thanks a tonne @natmokval !

natmokval commented:
thanks @MarcoGorelli, for helping me with this PR!
